File: lessons/Sorting/4-merge-sort/method-2/merge_sort.cpp
#include <iostream>
using namespace std;

void display(int arr[], int len) {
  for (int i = 0; i < len; i++) {
    cout << arr[i] << " ";
  }
}

void merge(int arr[], int start, int mid, int end) {
  // new array
  int mix[end - start];

  int i = start; // pointer for left arr
  int j = mid;   // pointer for right arr
  int k = 0;     // common ptr for both arr

  while (i < mid && j < end) {
    if (arr[i] < arr[j]) {
      mix[k] = arr[i];
      i++; // since left element is smaller put it in mix array and increment
           // pointer
    } else {
      mix[k] = arr[j];
      j++; // since right element is smaller put it in mix array and increment
           // pointer
    }
    k++; // increment mix array pointer
  }

  // if left array is not empty
  while (i < mid) {
    mix[k] = arr[i];
    i++;
    k++;
  }

  // if right array is not empty
  while (j < end) {
    mix[k] = arr[j];
    j++;
    k++;
  }

  // copy mix array to original array
  for (int l = 0; l < end - start; l++) {
    arr[start + l] = mix[l]; // start + l will give the correct index that needs
                             // to be updated
  }
}

void mergesort(int arr[], int start, int end) {

  // Base condition - when array only as one element
  if (end - start == 1) {
    return;
  }

  // Divide the array into two parts
  int mid = start + (end - start) / 2;

  // Recursively sort the two parts
  mergesort(arr, start, mid);
  mergesort(arr, mid, end);
  // Merge the two sorted parts accordingly
  merge(arr, start, mid, end);
}

int main() {

  int arr[] = {4, 2, 1, 3, 5, 6, 7, 8, 9, 10};
  int len = sizeof(arr) / sizeof(int);
  cout << "Input: ";
  display(arr, len);

  mergesort(arr, 0, len);

  cout << endl << "Output: ";
  display(arr, len);
  cout << endl;
  return 0;
}

File: archive/lessons/Sorting/merge-sort/method-1/merge_sort.java
import java.util.Arrays;

class MergeSort {
  public static void main(String[] args) {
    int[] arr = {5, 8, -1, 0, 4, 2, 6, 3, 1, 7};

    System.out.println("\nUnsorted array: ");
    for (int i = 0; i < arr.length; i++) {
      System.out.print(arr[i] + " ");
    }

    int[] sorted = mergesort(arr);

    System.out.println("\nSorted array: ");
    for (int i = 0; i < sorted.length; i++) {
      System.out.print(sorted[i] + " ");
    }
    System.out.println();
  }

  // Merge sort function
  static int[] mergesort(int[] arr) {
    // Base condition - if array has only one element
    if (arr.length <= 1) {
      return arr;
    }

    int mid = arr.length / 2;

    // Sort left half
    int[] left = mergesort(Arrays.copyOfRange(arr, 0, mid));

    // Sort right half
    int[] right = mergesort(Arrays.copyOfRange(arr, mid, arr.length));

    // Merge left and right halves and return
    return merge(left, right);
  }

  // Merge function
  private static int[] merge(int[] left, int[] right) {
    int[] result = new int[left.length + right.length];

    int i = 0; // pointer for left array
    int j = 0; // pointer for right array
    int k = 0; // pointer for result array

    while (i < left.length && j < right.length) {
      if (left[i] <= right[j]) {
        result[k] = left[i];
        i++;
      } else {
        result[k] = right[j];
        j++;
      }
      k++;
    }

    // Copy remaining elements of left array
    while (i < left.length) {
      result[k] = left[i];
      i++;
      k++;
    }

    // Copy remaining elements of right array
    while (j < right.length) {
      result[k] = right[j];
      j++;
      k++;
    }

    return result;
  }
}
